src/xarray_regrid/methods/conservative.py

```diff
@@ -167,58 +168,32 @@ def conservative_regrid_dataset(
 
 def apply_weights(
     da: xr.DataArray,
-    weights: xr.DataArray,
-    coord: Hashable,
-    valid_frac: xr.DataArray,
+    weights: dict[Hashable, xr.DataArray],
     skipna: bool,
-) -> tuple[xr.DataArray, xr.DataArray]:
-    """Apply the weights to convert data to the new coordinates."""
-    coord_map = {f"target_{coord}": coord}
-    weights_norm = weights.copy()
+    nan_threshold: float,
+) -> xr.DataArray:
+    """Apply the weights over all regridding dimensions simultaneously with `xr.dot`."""
+    coords = list(weights.keys())
+    weight_arrays = list(weights.values())
 
     if skipna:
-        notnull = da.notnull()
-        # Renormalize the weights along this dim by the accumulated valid_frac
-        # along previous dimensions
-        if valid_frac.name != EMPTY_DA_NAME:
-            weights_norm = weights * (valid_frac / valid_frac.mean(dim=[coord])).fillna(
-                0
-            )
-
-    da_reduced: xr.DataArray = xr.dot(
-        da.fillna(0), weights_norm, dim=[coord], optimize=True
+        valid_frac = xr.dot(
```

Comment:
No need to track an intermediate `valid_frac` array, now we just compute it all at once with the optimizations available in einsum. Much better!